apps/prairielearn/src/lib/question-render.ts

```diff
@@ -396,66 +395,93 @@ function buildLocals({
 export async function getAndRenderVariant(
   variant_id: string | null,
   variant_seed: string | null,
-  locals: Record<string, any>,
+  locals: {
+    urlPrefix: string;
+    course: Course;
+    question: Question;
+    user: User;
+    authn_user: User;
+    course_instance?: CourseInstance;
+    course_instance_id?: string;
+    assessment?: Assessment;
+    assessment_instance?: AssessmentInstance;
+    assessment_question?: AssessmentQuestion;
+    group_config?: GroupConfig;
+    group_role_permissions?: { can_view: boolean; can_submit: boolean };
+    instance_question?: InstanceQuestionWithAllowGrade;
+    authz_data?: Record<string, any>;
+    authz_result?: Record<string, any>;
+    client_fingerprint_id?: string | null;
+    manualGradingInterface?: boolean;
+  },
   options?: {
     urlOverrides?: Partial<QuestionUrls>;
   },
 ) {
-  locals.question_course = await getQuestionCourse(locals.question, locals.course);
-  locals.question_is_shared = await sqldb.queryRow(
+  // We write a fair amount of unstructured data back into locals,
+  // so we'll cast it to `any` once so we don't have to do it every time.
+  const resultLocals = locals as any;
```

Comment:
👀 this was my attempt to make reads safe while not forcing us to carefully type writes. I don't want to spend much time optimizing writes at this point since I think a beter long-term object is to explicitly return data rather than writing into this huge mutable shared object.